src/OpenTelemetry.Instrumentation.Runtime/description.md

```diff
@@ -0,0 +1,198 @@
+# Runtime metrics description
+
+Metrics name are prefixed with the `process.runtime.dotnet.` namespace, following
+the general guidance for runtime metrics in the
+[specs](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/runtime-environment-metrics.md#runtime-environment-specific-metrics---processruntimeenvironment).
+Instrument Units [should follow](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/README.md#instrument-units)
+the Unified Code for Units of Measure.
+
+## GC related metrics
+
+The metrics in this section can be enabled by setting the
+`RuntimeMetricsOptions.IsGcEnabled` switch.
+
+| Name                                          | Description              | Units     | Instrument Type   | Value Type | Attribute Key(s) | Attribute Values |
+|-----------------------------------------------|--------------------------|-----------|-------------------|------------|------------------|------------------|
+| process.runtime.dotnet.**gc.totalmemorysize** | Total allocated bytes    | `By`      | ObservableGauge   | `Int64`    |                  |                  |
```

Comment:
I think many of the `ObservableGauge` here should be `ObservableUpDownCounter` once .NET has the support for up-down counters.

Reply:
We'll probably need to to get more precise on this one. Some different possible interpretations:
- Total amount of memory ever allocated from the GC since process start
- Total amount of memory commited by the GC currently
- Total amount of memory for live .NET objects + memory for free objects (objects that existed in the past and were collected but the memory hasn't yet been reused)
- Total amount of memory for live .NET objects only
- Total amount of memory for rooted .NET objects (objects that would still be live after a full GC)

Reply:
Some raw notes from our discussion:

process.runtime.dotnet.gc.heapsize: live object (not necessarily hold by the user objects/threads, could be hold by the GC itself) + fragmentation

all the gen adding up should <= committed (since GC commits at the granularity of page, and it normally would ask for multiple pages instead of just one for caching purpose)

neither of the above includes GC's own memory usage

github.com/Maoni0/mem-doc/blob/master/doc/.NETMemoryPerformanceAnalysis.md

GC.GetGenerationSize(n) is not providing the transactional behavior since we have to call it multiple times for different gen, it is okay as long as we mention it in the description/doc

remove gc.totalmemorysize since it can be covered by adding up different gens